src/lib/panorama.ts

```diff
@@ -259,147 +267,187 @@ function getTileUrl(pano: Panorama, x: number, y: number, zoom: number) {
+
+				// Often, two panoramas will load after each other with the same zoom level and cropWidth/cropHeight.
+				// If this is the case, we don't need to recalculate the SphereGeometry, as we've already done that work.
+				const key = `${startX}_${widthX}_${startY}_${widthY}`;
+				if (!geometryCache[key]) {
+					geometryCache[key] = new THREE.SphereGeometry(999, 32, 16, -startX, -widthX, startY, widthY);
+				}
+
+				const geometry = geometryCache[key];
+				const loader = new THREE.TextureLoader(loadingManager);
+				const texture = loader.load(getTileUrl(meta.pano, x, y, zoom));
+				texture.colorSpace = THREE.SRGBColorSpace;
+				texture.matrixAutoUpdate = false;
+				texture.matrix.set(scaleX, 0, 0, 0, scaleY, shiftY, 0, 0, 1);
+				const material = new THREE.MeshBasicMaterial({
+					map: texture,
+					side: THREE.FrontSide,
+					transparent: true,
+					depthWrite: false,
+					opacity: 0,
+				});
+				const tile = new THREE.Mesh(geometry, material);
+				tile.renderOrder = RenderOrder.NEW_PANO;
+				panorama.add(tile);
+			}
+		}
 	});
 }
 
-// function renderErrorPano(scene: THREE.Scene, panoId: string) {
-// 	const canvas = document.createElement('canvas');
-// 	const ctx = canvas.getContext('2d')!;
-// 	canvas.width = 1024;
-// 	canvas.height = 1024;
-// 	ctx.fillStyle = '#be0039';
-// 	ctx.fillRect(0, 0, canvas.width, canvas.height);
-// 	ctx.lineWidth = 16;
-// 	ctx.strokeStyle = '#6d001a';
-// 	ctx.strokeRect(0, 0, canvas.width, canvas.height);
-// 	if (panoId !== '') {
-// 		ctx.fillStyle = 'white';
-// 		ctx.font = 'bold 60px sans-serif';
-// 		ctx.textAlign = 'center';
-// 		ctx.textBaseline = 'middle';
-// 		ctx.fillText(`Failed to load panorama.`, canvas.width / 2, canvas.height / 2 - 40);
-// 		ctx.font = 'bold 30px sans-serif';
-// 		ctx.fillText(`Send a screenshot to Mika if you see this.`, canvas.width / 2, canvas.height / 2 + 20);
-// 		ctx.font = 'bold 15px sans-serif';
-// 		ctx.fillText(`PanoID: ${panoId}`, canvas.width / 2, canvas.height / 2 + 55);
-// 	}
-
-// 	const texture = new THREE.CanvasTexture(canvas);
-// 	texture.colorSpace = THREE.SRGBColorSpace;
-// 	const material = new THREE.MeshBasicMaterial({
-// 		map: texture,
-// 		side: THREE.DoubleSide,
-// 		transparent: true,
-// 		depthWrite: false,
-// 		opacity: 0,
-// 	});
-// 	const geometry = new THREE.BoxGeometry(999, 999, 999);
-// 	geometry.scale(1, 1, -1);
-// 	const materials = [material, material, material, material, material, material];
-// 	const panorama = new THREE.Mesh(geometry, materials);
-// 	panorama.renderOrder = 1;
-
-// 	newPano = panorama;
-// 	scene.add(panorama);
-// 	panoFadeTime = performance.now();
-// }
+const errCanvas = document.createElement('canvas');
+const errContext = errCanvas.getContext('2d')!;
+errCanvas.width = 1024;
+errCanvas.height = 1024;
+
+const errTexture = new THREE.CanvasTexture(errCanvas);
+errTexture.colorSpace = THREE.SRGBColorSpace;
+
+const errorPanoGeometry = new THREE.BoxGeometry(999, 999, 999);
```

Comment:
Either rename this to `errPanoGeometry` or rename the other variables to `errorXYZ`, for consistency.